### PR TITLE
extract case update code to separate class

### PR DESCRIPTION
**Pull request by snopoke**
@benrudolph

Just moving code around. Extracting the code that updates a case into its own class.

cc @millerdev

### CODE REVIEW

File: corehq/ex-submodules/casexml/apps/case/update_strategy.py [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+from couchdbkit import BadValueError
+import sys
+from casexml.apps.case import const
+from casexml.apps.case.exceptions import ReconciliationError, MissingServerDate, UsesReferrals
+from casexml.apps.case.models import CommCareCase
+from casexml.apps.case.util import primary_actions
+from casexml.apps.case.xml.parser import KNOWN_PROPERTIES
+from django.utils.translation import ugettext as _
+from corehq.util.couch_helpers import CouchAttachmentsBuilder
+from corehq.util.dates import iso_string_to_datetime
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/case/update_strategy.py]: F401 'iso_string_to_datetime' imported but unused

File: corehq/ex-submodules/casexml/apps/case/update_strategy.py [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+from corehq.util.soft_assert import soft_assert
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/case/update_strategy.py]: F401 'soft_assert' imported but unused

File: corehq/ex-submodules/casexml/apps/case/update_strategy.py [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+    @classmethod
+    def case_from_case_update(cls, case_update, xformdoc):
+        """
+        Create a case object from a case update object.
+        """
+        assert not is_deprecation(xformdoc)  # you should never be able to create a case from a deleted update
+        case = CommCareCase()
+        case._id = case_update.id
+        case.modified_on = parsing.string_to_utc_datetime(case_update.modified_on_str) \
+                            if case_update.modified_on_str else datetime.utcnow()
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/case/update_strategy.py]: E127 continuation line over-indented for visual indent

File: corehq/ex-submodules/casexml/apps/case/update_strategy.py [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+        self.soft_rebuild_case(strict=False, xforms=local_forms)
+
+        if case_update.version:
+            self.case.version = case_update.version
+
+    def reset_case_state(self):
+        """
+        Clear known case properties, and all dynamic properties
+        """
+        dynamic_properties = set([k for action in self.case.actions for k in action.updated_unknown_properties.keys()])
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/case/update_strategy.py]: E501 line too long (119 > 115 characters)

**Comment by snopoke** [comment]
:white_check_mark: 

cc @dannyroberts @czue

**Comment by czue** [comment]
didn't review closely but looks good! trusting in code move + tests

**Comment by benrudolph** [comment]
:+1:

**Pull request merged**